sdk/python/v1beta1/kubeflow/katib/api/katib_client.py

```diff
@@ -302,137 +424,347 @@ def tune(
+                persistent_volume_claim=models.V1PersistentVolumeClaimVolumeSource(
+                    claim_name=name
+                ),
+            )
+
+            # create worker pod spec
+            worker_pod_template_spec = training_utils.get_pod_template_spec(
+                containers=[container_spec],
+                volumes=[storage_initializer_volume],
+            )
+
+            # create master pod spec
+            master_pod_template_spec = training_utils.get_pod_template_spec(
+                containers=[container_spec],
+                init_containers=[init_container_spec],
+                volumes=[storage_initializer_volume],
+            )
+
+            # Create pytorchjob.
+            pytorchjob = training_models.KubeflowOrgV1PyTorchJob(
+                api_version="kubeflow.org/v1",
+                kind="PyTorchJob",
+                spec=training_models.KubeflowOrgV1PyTorchJobSpec(
+                    run_policy=training_models.KubeflowOrgV1RunPolicy(
+                        clean_pod_policy=None
                     ),
+                    pytorch_replica_specs={},
+                ),
+            )
+
+            if resources_per_trial.num_procs_per_worker:
+                pytorchjob.spec.nproc_per_node = str(
+                    resources_per_trial.num_procs_per_worker
                 )
-            ),
-        )
 
-        # Create Trial template.
-        trial_template = models.V1beta1TrialTemplate(
-            primary_container_name=constants.DEFAULT_PRIMARY_CONTAINER_NAME,
-            retain=retain_trials,
-            trial_parameters=trial_params,
-            trial_spec=trial_spec,
-        )
+            pytorchjob.spec.pytorch_replica_specs["Master"] = (
+                training_models.KubeflowOrgV1ReplicaSpec(
+                    replicas=1,
+                    template=master_pod_template_spec,
+                )
+            )
+
+            if resources_per_trial.num_workers > 1:
+                pytorchjob.spec.pytorch_replica_specs["Worker"] = (
+                    training_models.KubeflowOrgV1ReplicaSpec(
+                        replicas=resources_per_trial.num_workers - 1,
+                        template=worker_pod_template_spec,
+                    )
+                )
+
+            # Add metrics collector to the Katib Experiment.
+            # Specify metrics format for the collector, for example: 'train_loss':0.846
+            experiment.spec.metrics_collector_spec = models.V1beta1MetricsCollectorSpec(
+                collector=models.V1beta1CollectorSpec(
+                    kind=metrics_collector_config["kind"]
+                ),
+                source=models.V1beta1SourceSpec(
+                    filter=models.V1beta1FilterSpec(
+                        metrics_format=[
+                            r"'([\w|-]+)'\s*:\s*([+-]?\d*(\.\d+)?([Ee][+-]?\d+)?)",
+                        ]
+                    )
+                ),
+            )
```

Comment:
Since we support only StdOut metrics collector for Trainer right now, I think you can remove this part.

```suggestion
            experiment.spec.metrics_collector_spec = models.V1beta1MetricsCollectorSpec(
                source=models.V1beta1SourceSpec(
                    filter=models.V1beta1FilterSpec(
                        metrics_format=[
                            r"'([\w|-]+)'\s*:\s*([+-]?\d*(\.\d+)?([Ee][+-]?\d+)?)",
                        ]
                    )
                ),
            )
```

Reply:
Thank you for your careful review! I have updated accordingly.

Reply:
@andreyvelich I have removed this part and moved the following code before the two options to make it clearer (it was previously inside the custom objective function option):
https://github.com/kubeflow/katib/blob/c2df96757ce2fb39b5debf60ad7ed1873ce31ae5/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py#L418
Do you think this makes sense?

Reply:
I think, that movement is not necessary since you override it for LLM option here: https://github.com/kubeflow/katib/blob/c2df96757ce2fb39b5debf60ad7ed1873ce31ae5/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py#L697-L705.
So, I think, you can move that code just after this check: https://github.com/kubeflow/katib/blob/c2df96757ce2fb39b5debf60ad7ed1873ce31ae5/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py#L426

Reply:
Got it! I will move it after this check.